packages/Search/src/details/DTK_DetailsDistributedSearchTreeImpl.hpp

```diff
@@ -198,36 +206,96 @@ DistributedSearchTreeImpl<DeviceType>::sendAcrossNetwork(
-                                             point[1] - eps[1],
-                                             point[2] - eps[2],
-                                         }},
-                                         {{
-                                             point[0] + eps[0],
-                                             point[1] + eps[1],
-                                             point[2] + eps[2],
-                                         }}};
-                              overlap_queries( q ) = Details::Overlap( box );
+                          KOKKOS_LAMBDA( int i ) {
+                              for ( int j = offset( i ); j < offset( i + 1 );
+                                    ++j )
+                                  farthest_distances( i ) = KokkosHelpers::max(
+                                      farthest_distances( i ), distances( j ) );
                           } );
     Kokkos::fence();
 
-    top_tree.query( overlap_queries, indices, offset );
+    // Identify what ranks may have leaves that are within that distance.
+    Kokkos::View<Details::Within *, DeviceType> within_queries( "queries",
+                                                                n_queries );
+    Kokkos::parallel_for(
+        DTK_MARK_REGION( "bottom_trees_within_that_distance" ),
+        Kokkos::RangePolicy<ExecutionSpace>( 0, n_queries ),
+        KOKKOS_LAMBDA( int i ) {
+            within_queries( i ) = Details::within( queries( i )._geometry,
+                                                   farthest_distances( i ) );
+        } );
+    Kokkos::fence();
+
+    top_tree.query( within_queries, indices, offset );
+    // NOTE: in principle, we could perform within queries on the bottom_tree
+    // rather than nearest queries.
+}
+
+template <typename DeviceType>
+template <typename Query>
+void DistributedSearchTreeImpl<DeviceType>::deviseStrategy(
+    Kokkos::View<Query *, DeviceType> queries,
+    DistributedSearchTree<DeviceType> const &tree,
+    Kokkos::View<int *, DeviceType> &indices,
+    Kokkos::View<int *, DeviceType> &offset,
+    Kokkos::View<double *, DeviceType> & )
+{
+    auto const &top_tree = tree._top_tree;
+    auto const &bottom_tree_sizes = tree._bottom_tree_sizes;
+
+    top_tree.query( queries, indices, offset );
+
+    auto const n_queries = queries.extent( 0 );
+    Kokkos::View<int *, DeviceType> _offset( offset.label(), n_queries + 1 );
+    Kokkos::deep_copy( _offset, 0 );
+    Kokkos::parallel_for(
+        DTK_MARK_REGION( "bottom_trees_with_required_cumulated_leaves_count" ),
+        Kokkos::RangePolicy<ExecutionSpace>( 0, n_queries ),
+        KOKKOS_LAMBDA( int i ) {
+            int leaves_count = 0;
+            for ( int j = offset( i ); j < offset( i + 1 ); ++j )
+            {
+                leaves_count += bottom_tree_sizes[indices( j )];
+                ++_offset( i );
+                if ( leaves_count >= queries( i )._k )
```

Comment:
What does this condition mean?

Reply:
That the ranks we will forward queries to have a cumulated total number of leaves big enough to find all k neighbors queried for.